kAFL-Fuzzer/common/qemu.py

```diff
@@ -425,55 +410,99 @@ def shutdown(self):
+                f.truncate(self.agent_size)
+                f.seek(0)
+                f.write(bin_)
+        except FileExistsError:
+            pass
 
     def start(self):
 
         if self.exiting:
             return False
 
-        self.persistent_runs = 0
-        self.handshake_stage_1 = True
-        self.handshake_stage_2 = True
+        if self.process is not None:
+            print_fail("Trying to launch Qemu second time?")
+            raise Exception
 
-        if self.qemu_id == "0" or self.qemu_id == "1337": ## 1337 is debug instance!
-            log_qemu("Launching virtual machine...CMD:\n" + ' '.join(self.cmd), self.qemu_id)
-        else:
-            log_qemu("Launching virtual machine...", self.qemu_id)
+        max_attempts = 10
+        for attempt in range(max_attempts):
+            time.sleep(attempt + 10 * random.random())
+
+            self.persistent_runs = 0
+            self.handshake_stage_1 = True
+            self.handshake_stage_2 = True
+
+            if self.qemu_id == "0" or self.qemu_id == "1337":   # 1337 is debug instance!
+                log_qemu("Launching virtual machine...CMD:\n" + ' '.join(self.cmd), self.qemu_id)
+            else:
+                log_qemu("Launching virtual machine...", self.qemu_id)
 
+            self.kafl_shm_f = os.open(self.bitmap_filename, os.O_RDWR | os.O_SYNC | os.O_CREAT)
+            self.fs_shm_f = os.open(self.payload_filename, os.O_RDWR | os.O_SYNC | os.O_CREAT)
 
-        # Launch Qemu. stderr to stdout, stdout is logged on VM exit
-        # os.setpgrp() prevents signals from being propagated to Qemu, instead allowing an
-        # organized shutdown via async_exit()
-        self.process = subprocess.Popen(self.cmd,
+            os.ftruncate(self.kafl_shm_f, self.bitmap_size)
+            os.ftruncate(self.fs_shm_f, self.payload_size)
+
+            # Launch Qemu. stderr to stdout, stdout is logged on VM exit
+            # os.setpgrp() prevents signals from being propagated to Qemu, instead allowing an
+            # organized shutdown via async_exit()
+            self.process = subprocess.Popen(
+                self.cmd,
                 preexec_fn=os.setpgrp,
                 stdin=subprocess.PIPE,
                 stdout=subprocess.PIPE,
-                stderr=subprocess.STDOUT)
+                stderr=subprocess.STDOUT
+            )
+            log_qemu("... launched PID %d" % self.process.pid, self.qemu_id)
 
-        try:
-            self.__qemu_connect()
-            self.__qemu_handshake()
-        except (OSError, BrokenPipeError) as e:
-            if not self.exiting:
-                print_fail("Failed to launch Qemu, please see logs. Error: " + str(e))
-                log_qemu("Fatal error: Failed to launch Qemu: " + str(e), self.qemu_id)
+            try:
+                self.__qemu_connect()
+                self.__qemu_handshake()
+                break
+            except socket.timeout:
```

Comment:
I think we can get other errors here if Qemu dies shortly after launch. Don't remember the failure cases, perhaps invalid -kernel image or something resulting in triple fault would lead to exit.